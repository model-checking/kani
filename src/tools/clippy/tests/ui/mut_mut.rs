// aux-build:macro_rules.rs

#![allow(unused, clippy::no_effect, clippy::unnecessary_operation)]
#![warn(clippy::mut_mut)]

#[macro_use]
extern crate macro_rules;

fn fun(x: &mut &mut u32) -> bool {
    **x > 0
}

fn less_fun(x: *mut *mut u32) {
    let y = x;
}

macro_rules! mut_ptr {
    ($p:expr) => {
        &mut $p
    };
}

#[allow(unused_mut, unused_variables)]
fn main() {
    let mut x = &mut &mut 1u32;
    {
        let mut y = &mut x;
    }

    if fun(x) {
        let y: &mut &mut u32 = &mut &mut 2;
        **y + **x;
    }

    if fun(x) {
        let y: &mut &mut &mut u32 = &mut &mut &mut 2;
        ***y + **x;
    }

    let mut z = mut_ptr!(&mut 3u32);
}

fn issue939() {
    let array = [5, 6, 7, 8, 9];
    let mut args = array.iter().skip(2);
    for &arg in &mut args {
        println!("{}", arg);
    }

    let args = &mut args;
    for arg in args {
        println!(":{}", arg);
    }
}

fn issue6922() {
    // do not lint from an external macro
    mut_mut!();
}
