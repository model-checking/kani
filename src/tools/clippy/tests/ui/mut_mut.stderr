error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:9:11
   |
LL | fn fun(x: &mut &mut u32) -> bool {
   |           ^^^^^^^^^^^^^
   |
   = note: `-D clippy::mut-mut` implied by `-D warnings`

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:25:17
   |
LL |     let mut x = &mut &mut 1u32;
   |                 ^^^^^^^^^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:19:9
   |
LL |         &mut $p
   |         ^^^^^^^
...
LL |     let mut z = mut_ptr!(&mut 3u32);
   |                 ------------------- in this macro invocation
   |
   = note: this error originates in the macro `mut_ptr` (in Nightly builds, run with -Z macro-backtrace for more info)

error: this expression mutably borrows a mutable reference. Consider reborrowing
  --> $DIR/mut_mut.rs:27:21
   |
LL |         let mut y = &mut x;
   |                     ^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:31:32
   |
LL |         let y: &mut &mut u32 = &mut &mut 2;
   |                                ^^^^^^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:31:16
   |
LL |         let y: &mut &mut u32 = &mut &mut 2;
   |                ^^^^^^^^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:36:37
   |
LL |         let y: &mut &mut &mut u32 = &mut &mut &mut 2;
   |                                     ^^^^^^^^^^^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:36:16
   |
LL |         let y: &mut &mut &mut u32 = &mut &mut &mut 2;
   |                ^^^^^^^^^^^^^^^^^^

error: generally you want to avoid `&mut &mut _` if possible
  --> $DIR/mut_mut.rs:36:21
   |
LL |         let y: &mut &mut &mut u32 = &mut &mut &mut 2;
   |                     ^^^^^^^^^^^^^

error: aborting due to 9 previous errors

